File: functions/uc3-gcp/uc3-aggregate/src/main/java/uc3/Aggregate.java
package uc3;

import com.google.auth.oauth2.GoogleCredentials;
import com.google.cloud.firestore.DocumentReference;
import com.google.cloud.firestore.DocumentSnapshot;
import com.google.cloud.firestore.Firestore;
import com.google.cloud.firestore.FirestoreOptions;
import com.google.cloud.functions.*;
import com.google.events.cloud.pubsub.v1.Message;
import com.google.gson.Gson;
import titan.ccp.model.records.ActivePowerRecord;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

public class Aggregate implements HttpFunction, BackgroundFunction<Message> {
    private Firestore db;
    private final String collection;
    private final int windowSize;
    private final int interval;

    public Aggregate() {
        // get environment variable COLLECTION
        this.collection = System.getenv("SELECT_COLLECTION");
        // get environment variable PROJECT_ID
        String projectId = System.getenv("GOOGLE_CLOUD_PROJECT");
        // get environment variable WINDOW_SIZE
        this.windowSize = Integer.parseInt(System.getenv("WINDOW_SIZE"));
        // get environment variable INTERVAL
        this.interval = Integer.parseInt(System.getenv("INTERVAL"));

        // create Firestore instance
        try {
            FirestoreOptions firestoreOptions = FirestoreOptions.getDefaultInstance().toBuilder()
                    .setProjectId(projectId)
                    .setCredentials(GoogleCredentials.getApplicationDefault())
                    .build();
            this.db = firestoreOptions.getService();
        } catch (Error | IOException ignored) {

        }
    }

    private void logFirestore(String operation, int count) {
        System.out.printf("FIRESTORE OP %s: %d\n", operation, count);
    }

    @Override
    public void service(HttpRequest request, HttpResponse response)
            throws IOException {
        // step 1: parse JSON
        Gson gson = new Gson();
        Map map = gson.fromJson(request.getReader().lines().collect(Collectors.joining(System.lineSeparator())),
                Map.class);

        this.handle(map);
    }

    @Override
    public void accept(Message message, Context context) {
        if (message.getData() == null) {
            System.out.println("No message provided");
            return;
        }

        try {
            byte[] bytes = Base64.getDecoder().decode(message.getData().getBytes(StandardCharsets.UTF_8));
            ActivePowerRecord apr = ActivePowerRecord.fromByteBuffer(ByteBuffer.wrap(bytes));
            System.out.println("Received message: " + apr.toString());
            this.handle(new HashMap<String, Object>() {
                {
                    put("timestamp", apr.getTimestamp());
                    put("identifier", apr.getIdentifier());
                    put("valueInW", apr.getValueInW());
                }
            });
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void handle(Map map) throws IOException {

        // select the attrbute we want
        String id = (String) map.get("identifier");
        double i = (double) map.get("valueInW");

        // iterate over windows

        // check the firestore collection
        // get current time
        long now = System.currentTimeMillis() / 1000;

        // iterate over all the entries (= open windows)
        // add values to open windows
        // close windows when they are deadlined
        // create new windows if required
        // there should be 2*windowSize/interval - 1 windows open at all times
        // offsets -windowsize+interval, -windowsize+2interval, ..., 0, interval,
        // 2interval, ..., windowsize-interval
        //
        // - -|- - -
        // -|- - - -
        // |- - - - -|
        // | - - - -|-
        // | - - -|- -
        // | - -|- - -
        // | -|- - - -
        // | |- - - - -
        // 0 1|2 3 4 5 6 7 8 9 1 1
        //
        // Aggregation uc1 are based on
        // https://guava.dev/releases/22.0/api/docs/com/google/common/math/StatsAccumulator.html
        // * count
        // * min
        // * max
        // * mean
        // * populationStandardDeviation
        //

        for (long t = 0; t < windowSize; t += interval) {
            // get the timestamp
            String wid = String.format("%s_%d", id, t);

            long T = now - (now % windowSize) + t;

            // System.out.println("Getting document " + wid);
            DocumentReference d = db.collection(collection).document(wid);

            DocumentSnapshot doc = null;

            try {
                doc = d.get().get();
            } catch (Error | ExecutionException | InterruptedException ignored) {
            } finally {
                logFirestore("READ", 1);
            }

            if ((doc == null) || (!doc.exists())) {
                // create a new document

                try {
                    d.set(new HashMap<String, Object>() {
                        {
                            put("T", T);
                            put("Count", (int) 1);
                            put("Min", (double) i);
                            put("Max", (double) i);
                            put("Mean", (double) i);
                            put("SD", (double) 0.0);
                            put("SD_M2", (double) 0.0);
                        }
                    }).get();
                } catch (Error | ExecutionException | InterruptedException ignored) {
                }
                logFirestore("WRITE", 1);

                try {
                    doc = d.get().get();
                } catch (Error | ExecutionException | InterruptedException e) {
                    e.printStackTrace();
                    throw new IOException(e);
                } finally {
                    logFirestore("READ", 1);
                }
                continue;
            }

            if ((doc.get("T") == null) || ((long) doc.get("T") == 0)) {
                throw new IOException(String.format("Failed to get window: T does not exist: %s", doc.get("T")));
            }

            if ((long) doc.get("T") < now - windowSize) {
                // window is dead

                // emit
                System.out.printf(
                        "{\"time\":%s, \"sensor\": %s, \"count\": %s, \"min\": %s ,\"max\": %s, \"mean\": %s, \"sd\": %s }\n",
                        T + t * interval, id, doc.get("Count"), doc.get("Min"), doc.get("Max"), doc.get("Mean"),
                        doc.get("SD"));

                // reset window
                d.set(new HashMap<String, Object>() {
                    {
                        put("T", T);
                        put("Count", 1);
                        put("Min", i);
                        put("Max", i);
                        put("Mean", i);
                        put("SD", 0.0);
                        put("SD_M2", 0.0);
                    }
                });
                logFirestore("WRITE", 1);
                continue;
            }

            int oldCount = 0;
            if (doc.get("Count", Integer.class) != null) {
                oldCount = doc.get("Count", Integer.class);
            }

            double oldMin = 0;
            if (doc.get("Min", Double.class) != null) {
                oldMin = doc.get("Max", Double.class);
            }

            double oldMax = 0;
            if (doc.get("Max", Double.class) != null) {
                oldMax = doc.get("Max", Double.class);
            }

            double oldMean = 0;
            if (doc.get("Mean", Double.class) != null) {
                oldMean = doc.get("Mean", Double.class);
            }

            double oldSDM2 = 0;
            if (doc.get("SD_M2", Double.class) != null) {
                oldSDM2 = doc.get("SD_M2", Double.class);
            }

            int newCount = oldCount + 1;
            double newMin = Math.min(oldMin, i);
            double newMax = Math.max(oldMax, i);
            double newMean = oldMean * oldCount + i / newCount;

            // https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
            // no idea if this is correctly implemented

            double delta = i - oldMean;
            double delta2 = i - (oldMean + delta / oldCount);
            double newSDM2 = oldSDM2 + delta * delta2;

            double newSD = (double) Math.sqrt(newSDM2 / newCount);

            d.set(new HashMap<String, Object>() {
                {
                    put("T", T);
                    put("Count", newCount);
                    put("Min", newMin);
                    put("Max", newMax);
                    put("Mean", newMean);
                    put("SD", newSD);
                    put("SD_M2", newSDM2);

                }
            });
            logFirestore("WRITE", 1);
        }
    }
}

File: functions/uc3-gcp/uc3-go/aggregate.go
package p

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"net/http"
	"os"
	"strconv"
	"time"

	"cloud.google.com/go/firestore"
)

var selectCollection string = os.Getenv("SELECT_COLLECTION")
var windowSize int64
var interval int64

var projectID string = os.Getenv("GOOGLE_CLOUD_PROJECT")

func init() {
	var err error
	windowSize, err = strconv.ParseInt(os.Getenv("WINDOW_SIZE"), 10, 64)
	if err != nil {
		log.Fatalf("Failed to parse WINDOW_SIZE: %v", err)
	}

	interval, err = strconv.ParseInt(os.Getenv("INTERVAL"), 10, 64)
	if err != nil {
		log.Fatalf("Failed to parse INTERVAL: %v", err)
	}
}

var f *firestore.Client

func logFirestore(t string, count int) {
	fmt.Printf("FIRESTORE OP %s: %d\n", t, count)
}

func aggregate(data []byte) error {
	// parse JSON
	c := &struct {
		ID        string  `json:"identifier"`
		Value     float64 `json:"valueInW"`
		Timestamp int64   `json:"timestamp"`
	}{}
	if err := json.Unmarshal(data, c); err != nil {
		log.Printf("Failed to parse data: %v", err)
		return err
	}

	// select the attrbute we want
	id := c.ID
	i := c.Value

	// iterate over windows
	// this is not threadsafe at all, maybe it should be?

	// check the firestore collection
	if f == nil {
		var err error
		f, err = firestore.NewClient(context.Background(), projectID)
		if err != nil {
			log.Printf("Failed to create client: %v", err)
			return err
		}
	}

	now := time.Now().UTC().Unix()
	// iterate over all the entries (= open windows)
	// add values to open windows
	// close windows when they are deadlined
	// create new windows if required
	// there should be 2*windowSize/interval - 1 windows open at all times
	// offsets -windowsize+interval, -windowsize+2interval, ..., 0, interval, 2interval, ..., windowsize-interval
	//
	// - -|- - -
	//   -|- - - -
	//    |- - - - -|
	//    |  - - - -|-
	//    |    - - -|- -
	//    |      - -|- - -
	//    |        -|- - - -
	//    |         |- - - - -
	// 0 1|2 3 4 5 6 7 8 9 1 1
	//
	// Aggregation uc1 are based on https://guava.dev/releases/22.0/api/docs/com/google/common/math/StatsAccumulator.html
	// * count
	// * min
	// * max
	// * mean
	// * populationStandardDeviation

	for t := int64(0); t < windowSize; t += interval {
		// get the window
		wid := fmt.Sprintf("%s_%d", id, t)
		T := now - (now % windowSize) + int64(t)

		d := f.Collection(selectCollection).Doc(wid)

		doc, err := d.Get(context.Background())
		logFirestore("READ", 1)
		if err != nil {
			// create if it does not exist

			_, err := d.Set(context.Background(), map[string]interface{}{
				"T":     T,
				"Count": 1,
				"Min":   float64(i),
				"Max":   float64(i),
				"Mean":  float64(i),
				"SD":    0.0,
				"SD_M2": 0.0,
			})
			logFirestore("WRITE", 1)

			if err != nil {
				log.Printf("Failed to create window: %v", err)
				return err
			}

			doc, err = d.Get(context.Background())
			logFirestore("READ", 1)
			if err != nil {
				log.Printf("Failed to get window: %v", err)
				return err
			}
			continue
		}

		if doc.Data()["T"] == nil || doc.Data()["T"].(int64) == 0 {
			log.Printf("Failed to get window: T does not exist: %v", doc.Data()["T"])
			return err
		}

		if doc.Data()["T"].(int64) < now-windowSize {
			// window is dead

			// emit
			log.Printf("{\"time\":%d, \"sensor\": %s, \"count\": %d, \"min\": %.2f ,\"max\": %.2f, \"mean\": %.2f, \"sd\": %.2f }\n", T+t*interval, id, doc.Data()["Count"], doc.Data()["Min"], doc.Data()["Max"], doc.Data()["Mean"], doc.Data()["SD"])

			// reset window
			_, err = d.Set(context.Background(), map[string]interface{}{
				"T":     T,
				"Count": 1,
				"Min":   float64(i),
				"Max":   float64(i),
				"Mean":  float64(i),
				"SD":    0.0,
				"SD_M2": 0.0,
			})
			logFirestore("WRITE", 1)
			if err != nil {
				log.Printf("Failed to reset window: %v", err)
				return err
			}
			continue
		}

		var oldCount int64
		var oldMin float64
		var oldMax float64
		var oldMean float64
		var oldSDM2 float64

		if doc.Data()["Count"] != nil {
			oldCount = doc.Data()["Count"].(int64)
		}

		if doc.Data()["Min"] != nil {
			oldMin = doc.Data()["Min"].(float64)
		}

		if doc.Data()["Max"] != nil {
			oldMax = doc.Data()["Max"].(float64)
		}

		if doc.Data()["Mean"] != nil {
			oldMean = doc.Data()["Mean"].(float64)
		}

		if doc.Data()["SD_M2"] != nil {
			oldSDM2 = doc.Data()["SD_M2"].(float64)
		}

		// update window
		newCount := oldCount + 1
		newMin := math.Min(oldMin, i)
		newMax := math.Max(oldMax, i)
		newMean := (oldMean*float64(oldCount) + i) / float64(newCount)

		// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
		// no idea if this is correctly implemented
		delta := i - oldMean
		delta2 := i - (oldMean + delta/float64(oldCount))
		newSDM2 := oldSDM2 + delta*delta2

		newSD := math.Sqrt(newSDM2 / float64(newCount))

		_, err = d.Set(context.Background(), map[string]interface{}{
			"T":     doc.Data()["T"].(int64),
			"Count": newCount,
			"Min":   newMin,
			"Max":   newMax,
			"Mean":  newMean,
			"SD":    newSD,
			"SD_M2": newSDM2,
		})
		logFirestore("WRITE", 1)
		if err != nil {
			log.Printf("Failed to create window: %v", err)
			return err
		}
	}

	return nil
}

// UC3AggregateHTTP consumes a Pub/Sub message.
func UC3AggregateHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}

	defer r.Body.Close()
	data, err := ioutil.ReadAll(r.Body)

	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to read body: %s", err.Error()), http.StatusInternalServerError)
	}

	err = aggregate(data)

	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to aggregate: %s", err.Error()), http.StatusInternalServerError)
	}
}
